src/PermissionTree.js

```diff
@@ -0,0 +1,167 @@
+/**
+ * PermissionTree class builds and handels the permission tree
+ * for the given models and methods
+ */
+module.exports = class PermissionTree {
+    constructor(models, remotes) {
+        /**
+         * Loopback data models
+         * @type {Array}
+         */
+        this.models = models;
+
+        /**
+         * Loopback remote methods
+         * @type {Array}
+         */
+        this.remotes = remotes;
+
+        /**
+        * Loopback access types
+        *
+        * @type {Array}
+        */
+        this.ACCESS_TYPES = ['READ', 'REPLICATE', 'WRITE', 'EXECUTE'];
+
+        /**
+         * Object to hold the builded user permission trees
+         * @type {Object}
+         */
+        this.userPermissionTrees = new Map();
+    }
+
+    _buildDefaultTree() {
+        const defaultTree = {};
+
+        Object.values(this.models).forEach((model) => {
+            if (model.shared) {
+                defaultTree[model.modelName] = {};
+                if (this.remotes._classes[model.modelName]) {
+                    const methods = this.remotes._classes[model.modelName]._methods;
+                    Object.values(methods).forEach((method) => {
+                        defaultTree[model.modelName][method.name] = {};
+
+                        this.ACCESS_TYPES.forEach((accessType) => {
+                            defaultTree[model.modelName][method.name][accessType] = false;
+                        });
+                    });
+                }
+            }
+        });
+
+        this.defaultTree = defaultTree;
+        return this.defaultTree;
+    }
+
+    getDefaultTree() {
+        if (this.defaultTree) {
+            return this.defaultTree;
+        }
+
+        return this._buildDefaultTree();
+    }
+
+
+    getUserPermissionTree(user) {
+        if (!user) {
+            throw new Error('No user was specified while trying to access the users permission tree!');
+        }
+
+        if (!this.userPermissionTrees.has(user.id)) {
+            // Deep clone
+            this.setUserPermissionTree(user, JSON.parse(JSON.stringify(this.getDefaultTree())));
```

Comment:
Not sure about this, interesting thought. I did id because there is a lot of nesting in the defaultTree function and I don't know what loopback is doing in the background when accessing the remotes.